README.md
# link_previewer_aad

A new Flutter package. Build container with image, title, description from web-site (link) will be provided.

[Original Package](https://pub.dev/packages/link_previewer_aad)

## Warning - This package might be removed if the changes gets approved on the original package.
## Getting Started

![alt text](https://github.com/dmlapteacru/flutter-link-previewer/blob/master/images/link_preview.PNG)

## Installing 
Use this package as a library
1. Depend on it

Add this to your package's pubspec.yaml file:


```yaml
dependencies:
  link_previewer_aad: ^1.0.0
```

2. Install it

You can install packages from the command line:
with Flutter:

```
$ flutter pub get
```

Alternatively, your editor might support flutter pub get. Check the docs for your editor to learn more.

3. Import it

Now in your Dart code, you can use:

```dart
import 'package:link_previewer_aad/link_previewer_aad.dart';
```

## Usage

```dart
Widget build() {
  return Column(
    mainAxisAlignment: MainAxisAlignment.center,
    children: <Widget>[
      LinkPreviewer(
        link: "https://www.linkedin.com/feed/",
        direction: ContentDirection.horizontal,
      ),
      LinkPreviewer(
        link: "https://www.linkedin.com/feed/",
        direction: ContentDirection.vertical,
      ),
    ],
  );
}
```

## Properties (Constructor props)
| PropName      | Value         | Definition  |
| ------------- |:-------------:| -----:|
| `link`          | `String` value (Ex: `"http://..."`) | Link to website will be shown as LinkPreviewer |
| `titleFontSize`      | `double` (default: `15`, but it's auto resizable depends on layout(parent) width or hight)    |  Title font size  |
| `bodyFontSize` | `double` (default: `14`, but it's auto resizable depends on layout(parent) width or hight)   |    Body font size |
| `backgroundColor` | `Color` (default: `Colors.white`) | Background color in case there will be not much text and will be whitespace |
| `borderColor` | `Color` (default: `Colors.deepOrangeAccent`) | Border color of the previewer |
| `borderRadius` | `double` (default: `3.0`) | Border radius |
| `direction` | `ContentDirection` from the package (default: `ContentDirection.horizontal`)                                            Options: 1. `ContentDirection.horizontal` 2. `ContentDirection.vertical` | Content direction (image -> title -> body) |

#### From version 1.1.0
| PropName  | Value | Definition |
| --------- |:-----:| ----------:|
| `placeholder` | `Widget` | Custom placeholder while data is loading |
| `defaultPlaceholderColor` | `Color` (default: `light-grey`) | If placeholder is not set, will be used the default one and its color can be set using this prop (empty container with background) |
| `showTitle` | `bool` (default: `true`) | Title can be hiden using this prop, if it set to false |
| `showBody` | `bool` (default: `true`) | Body can be hiden using this prop, if it set to false |

#### From version 1.2.0
| PropName  | Value | Definition |
| --------- |:-----:| ----------:|
| `bodyTextOverflow` | `TextOverflow` (default: `TextOverflow.ellipsis` or `TextOverflow.fade`) | Text overflow prop |
| `bodyMaxLines` | `int` (default: `depends on hight`) | Max Lines that will be shown in the body |


CHANGELOG.md
## [1.0.0]
### Init Null safety 



test/link_previewer_test.dart
/*
import 'package:flutter_test/flutter_test.dart';

import 'package:link_previewer_aad/link_previewer_aad.dart';
*/

void main() {}


lib/link_previewer_aad.dart
library link_previewer_aad;

import 'package:flutter/material.dart' hide Element;
import 'package:url_launcher/url_launcher.dart';

import 'parser/web_page_parser.dart';

part 'package:link_previewer_aad/content_direction.dart';
part 'package:link_previewer_aad/horizontal_link_view.dart';
part 'package:link_previewer_aad/vertical_link_preview.dart';

class LinkPreviewerAad extends StatefulWidget {
  LinkPreviewerAad({
    Key? key,
    required this.link,
    this.titleFontSize,
    this.bodyFontSize,
    this.backgroundColor = Colors.white,
    this.borderColor = Colors.deepOrangeAccent,
    this.titleTextColor = Colors.black,
    this.bodyTextColor = Colors.grey,
    this.defaultPlaceholderColor,
    this.borderRadius,
    this.placeholder,
    this.showTitle = true,
    this.showBody = true,
    this.direction = ContentDirection.horizontal,
    this.bodyTextOverflow,
    this.bodyMaxLines,
  }) : super(key: key);

  final String link;
  final double? titleFontSize;
  final double? bodyFontSize;
  final Color backgroundColor;
  final Color borderColor;
  final Color? defaultPlaceholderColor;
  final Color titleTextColor;
  final Color bodyTextColor;
  final double? borderRadius;
  final ContentDirection direction;
  final Widget? placeholder;
  final bool showTitle;
  final bool showBody;
  final TextOverflow? bodyTextOverflow;
  final int? bodyMaxLines;

  @override
  _LinkPreviewer createState() => _LinkPreviewer();
}

class _LinkPreviewer extends State<LinkPreviewerAad> {
  Map? _metaData;
  double? _height;
  String? _link;
  Color? _placeholderColor;
  bool _failedToLoadImage = false;

  @override
  void initState() {
    super.initState();
    _link = widget.link.trim();
    if (_link!.startsWith("https")) {
      _link = "http" + _link!.split("https")[1];
    }
    _placeholderColor = widget.defaultPlaceholderColor == null
        ? Color.fromRGBO(235, 235, 235, 1.0)
        : widget.defaultPlaceholderColor;
    _fetchData();
  }

  double _computeHeight(double screenHeight) {
    if (widget.direction == ContentDirection.horizontal) {
      return screenHeight * 0.12;
    } else {
      return screenHeight * 0.25;
    }
  }

  void _fetchData() {
    if (!isValidUrl(_link)) {
      throw Exception("Invalid link");
    } else {
      _getMetaData(_link);
    }
  }

  void _validateImageUri(uri) {
    precacheImage(NetworkImage(uri), context, onError: (e, stackTrace) {
      setState(() {
        _failedToLoadImage = true;
      });
    });
  }

  String? _getUriWithPrefix(uri) {
    return WebPageParser.addWWWPrefixIfNotExists(uri);
  }

  void _getMetaData(link) async {
    Map data = await WebPageParser.getData(link);
    if (data != null) {
      _validateImageUri(data['image']);
      setState(() {
        _metaData = data;
      });
    } else {
      setState(() {
        _metaData = null;
      });
    }
  }

  bool isValidUrl(link) {
    String regexSource =
        "^(https?)://[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]";
    final regex = RegExp(regexSource);
    final matches = regex.allMatches(link);
    for (Match match in matches) {
      if (match.start == 0 && match.end == link.length) {
        return true;
      }
    }
    return false;
  }

  void _launchURL(url) async {
    if (await canLaunch(url)) {
      await launch(url);
    } else {
      throw 'Could not launch $url';
    }
  }

  @override
  Widget build(BuildContext context) {
    _height = _computeHeight(MediaQuery.of(context).size.height);

    return _metaData == null
        ? widget.placeholder == null
            ? _buildPlaceHolder(_placeholderColor, _height)
            : widget.placeholder!
        : _buildLinkContainer();
  }

  Widget _buildPlaceHolder(Color? color, double? defaultHeight) {
    return Container(
      height: defaultHeight,
      child: LayoutBuilder(builder: (context, constraints) {
        var layoutWidth = constraints.biggest.width;
        var layoutHeight = constraints.biggest.height;

        return Container(
          decoration: new BoxDecoration(
            color: color,
            border: Border.all(
              color: widget.borderColor == null
                  ? widget.backgroundColor
                  : widget.borderColor,
              width: widget.borderColor == null ? 0.0 : 1.0,
            ),
            borderRadius: BorderRadius.all(Radius.circular(
                widget.borderRadius == null ? 3.0 : widget.borderRadius!)),
          ),
          width: layoutWidth,
          height: layoutHeight,
        );
      }),
    );
  }

  Widget _buildLinkContainer() {
    return Container(
      decoration: new BoxDecoration(
        color: widget.backgroundColor,
        border: Border.all(
          color: widget.borderColor == null
              ? widget.backgroundColor
              : widget.borderColor,
          width: widget.borderColor == null ? 0.0 : 1.0,
        ),
        borderRadius: BorderRadius.all(Radius.circular(
            widget.borderRadius == null ? 3.0 : widget.borderRadius!)),
      ),
      height: _height,
      child: _buildLinkView(
          _link,
          _metaData!['title'] == null ? "" : _metaData!['title'],
          _metaData!['description'] == null ? "" : _metaData!['description'],
          _metaData!['image'] == null ? "" : _metaData!['image'],
          _launchURL,
          widget.showTitle,
          widget.showBody,
          widget.borderRadius == null ? 3.0 : widget.borderRadius),
    );
  }

  Widget _buildLinkView(link, title, description, imageUri, onTap, showTitle,
      showBody, borderRadius) {
    if (widget.direction == ContentDirection.horizontal) {
      return HorizontalLinkView(
        url: link,
        title: title,
        description: description,
        imageUri: _failedToLoadImage == false
            ? imageUri
            : _getUriWithPrefix(imageUri)!,
        onTap: onTap,
        showTitle: showTitle,
        showBody: showBody,
        bodyTextOverflow: widget.bodyTextOverflow,
        bodyMaxLines: widget.bodyMaxLines,
        titleTextColor: widget.titleTextColor,
        bodyTextColor: widget.bodyTextColor,
        borderRadius: borderRadius,
      );
    } else {
      return VerticalLinkPreview(
        url: link,
        title: title,
        description: description,
        imageUri: _failedToLoadImage == false
            ? imageUri
            : _getUriWithPrefix(imageUri)!,
        onTap: onTap,
        showTitle: showTitle,
        showBody: showBody,
        bodyTextOverflow: widget.bodyTextOverflow,
        bodyMaxLines: widget.bodyMaxLines,
        titleTextColor: widget.titleTextColor,
        bodyTextColor: widget.bodyTextColor,
        borderRadius: borderRadius,
      );
    }
  }
}


lib/horizontal_link_view.dart
part of link_previewer_aad;

class HorizontalLinkView extends StatelessWidget {
  HorizontalLinkView({
    Key? key,
    required this.url,
    required this.title,
    required this.description,
    required this.imageUri,
    required this.onTap,
    this.titleFontSize,
    this.bodyFontSize,
    this.showTitle,
    this.showBody,
    this.bodyTextOverflow,
    this.bodyMaxLines,
    this.titleTextColor,
    this.bodyTextColor,
    this.borderRadius,
  }) : super(key: key);

  final String url;
  final String title;
  final String description;
  final String imageUri;
  final Function onTap;
  final double? titleFontSize;
  final double? bodyFontSize;
  final bool? showTitle;
  final bool? showBody;
  final TextOverflow? bodyTextOverflow;
  final int? bodyMaxLines;
  final Color? titleTextColor;
  final Color? bodyTextColor;
  final double? borderRadius;

  double computeTitleFontSize(double width) {
    double size = width * 0.13;
    if (size > 15) {
      size = 15;
    }
    return size;
  }

  int computeTitleLines(layoutHeight) {
    return layoutHeight >= 100 ? 2 : 1;
  }

  int computeBodyLines(double layoutHeight) {
    var lines = 1;
    if (layoutHeight > 40) {
      lines += (layoutHeight - 40.0) ~/ 15.0;
    }
    return lines;
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(builder: (context, constraints) {
      var layoutWidth = constraints.biggest.width;
      var layoutHeight = constraints.biggest.height;

      var _titleFontSize = titleFontSize == null
          ? computeTitleFontSize(layoutWidth)
          : titleFontSize;
      var _bodyFontSize = bodyFontSize == null
          ? computeTitleFontSize(layoutWidth) - 1
          : bodyFontSize;

      return InkWell(
        onTap: () => onTap(url),
        child: Row(
          children: <Widget>[
            Expanded(
              flex: 1,
              child: ClipRRect(
                borderRadius: BorderRadius.horizontal(
                    left: Radius.circular(borderRadius!)),
                child: imageUri == ""
                    ? Container(
                        color: Color.fromRGBO(235, 235, 235, 1.0),
                      )
                    : Container(
                        foregroundDecoration: BoxDecoration(
                          image: DecorationImage(
                              image: NetworkImage(imageUri), fit: BoxFit.cover),
                        ),
                      ),
              ),
            ),
            Expanded(
              flex: 3,
              child: Column(
                mainAxisAlignment: showBody == false
                    ? MainAxisAlignment.center
                    : MainAxisAlignment.start,
                children: <Widget>[
                  showTitle == false
                      ? Container()
                      : _buildTitleContainer(
                          _titleFontSize, computeTitleLines(layoutHeight)),
                  showBody == false
                      ? Container()
                      : _buildBodyContainer(
                          _bodyFontSize, computeBodyLines(layoutHeight))
                ],
              ),
            ),
          ],
        ),
      );
    });
  }

  Widget _buildTitleContainer(_titleFontSize, _maxLines) {
    return Padding(
      padding: const EdgeInsets.fromLTRB(4.0, 2.0, 3.0, 1.0),
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Container(
            alignment: Alignment(-1.0, -1.0),
            child: Text(
              title,
              style: TextStyle(
                  fontWeight: FontWeight.bold,
                  fontSize: _titleFontSize,
                  color: titleTextColor),
              overflow: TextOverflow.ellipsis,
              maxLines: _maxLines,
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildBodyContainer(_bodyFontSize, _maxLines) {
    return Expanded(
      flex: 2,
      child: Padding(
        padding: const EdgeInsets.fromLTRB(5.0, 2.0, 5.0, 0.0),
        child: Column(
          mainAxisAlignment: showTitle == false
              ? MainAxisAlignment.center
              : MainAxisAlignment.start,
          children: <Widget>[
            Expanded(
              child: Container(
                alignment: Alignment(-1.0, -1.0),
                child: Text(
                  description,
                  textAlign: TextAlign.left,
                  style:
                      TextStyle(fontSize: _bodyFontSize, color: bodyTextColor),
                  overflow: bodyTextOverflow == null
                      ? TextOverflow.ellipsis
                      : bodyTextOverflow,
                  maxLines: bodyMaxLines == null ? _maxLines : bodyMaxLines,
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/vertical_link_preview.dart
part of link_previewer_aad;

class VerticalLinkPreview extends StatelessWidget {
  VerticalLinkPreview({
    Key? key,
    required this.url,
    required this.title,
    required this.description,
    required this.imageUri,
    required this.onTap,
    this.titleFontSize,
    this.bodyFontSize,
    this.showTitle,
    this.showBody,
    this.bodyTextOverflow,
    this.bodyMaxLines,
    this.titleTextColor,
    this.bodyTextColor,
    this.borderRadius,
  })  : assert(imageUri != null),
        assert(title != null),
        assert(url != null),
        assert(description != null),
        assert(onTap != null),
        super(key: key);

  final String url;
  final String title;
  final String description;
  final String imageUri;
  final Function onTap;
  final double? titleFontSize;
  final double? bodyFontSize;
  final bool? showTitle;
  final bool? showBody;
  final TextOverflow? bodyTextOverflow;
  final int? bodyMaxLines;
  final Color? titleTextColor;
  final Color? bodyTextColor;
  final double? borderRadius;

  double computeTitleFontSize(double height) {
    double size = height * 0.13;
    if (size > 15) {
      size = 15;
    }
    return size;
  }

  int computeTitleLines(layoutHeight, layoutWidth) {
    return layoutHeight - layoutWidth < 50 ? 1 : 2;
  }

  int? computeBodyLines(layoutHeight) {
    return layoutHeight ~/ 60 == 0 ? 1 : layoutHeight ~/ 60;
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(builder: (context, constraints) {
      var layoutWidth = constraints.biggest.width;
      var layoutHeight = constraints.biggest.height;

      var _titleFontSize = titleFontSize == null
          ? computeTitleFontSize(layoutHeight)
          : titleFontSize;
      var _bodyFontSize = bodyFontSize == null
          ? computeTitleFontSize(layoutHeight) - 1
          : bodyFontSize;

      return InkWell(
          onTap: () => onTap(url),
          child: Column(
            children: <Widget>[
              Expanded(
                flex: 2,
                child: ClipRRect(
                  borderRadius: BorderRadius.vertical(
                      top: Radius.circular(borderRadius!)),
                  child: imageUri == ""
                      ? Container(
                          color: Color.fromRGBO(235, 235, 235, 1.0),
                        )
                      : Container(
                          foregroundDecoration: BoxDecoration(
                            image: DecorationImage(
                                image: NetworkImage(imageUri),
                                fit: layoutHeight >= layoutWidth
                                    ? BoxFit.cover
                                    : BoxFit.fitWidth),
                          ),
                        ),
                ),
              ),
              showTitle == false
                  ? Container()
                  : _buildTitleContainer(_titleFontSize,
                      computeTitleLines(layoutHeight, layoutWidth)),
              showBody == false
                  ? Container()
                  : _buildBodyContainer(
                      _bodyFontSize, computeBodyLines(layoutHeight)),
            ],
          ));
    });
  }

  Widget _buildTitleContainer(_titleFontSize, _maxLines) {
    return Padding(
      padding: const EdgeInsets.fromLTRB(4.0, 1.0, 3.0, 0.0),
      child: Container(
        alignment: Alignment(-1.0, -1.0),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              title,
              style: TextStyle(
                  fontWeight: FontWeight.bold,
                  fontSize: _titleFontSize,
                  color: titleTextColor),
              overflow: TextOverflow.ellipsis,
              maxLines: _maxLines,
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildBodyContainer(_bodyFontSize, _maxLines) {
    return Expanded(
      flex: 1,
      child: Padding(
        padding: const EdgeInsets.fromLTRB(5.0, 2.0, 5.0, 4.0),
        child: Container(
          alignment: Alignment(-1.0, -1.0),
          child: Text(
            description,
            style: TextStyle(fontSize: _bodyFontSize, color: bodyTextColor),
            overflow:
                bodyTextOverflow == null ? TextOverflow.fade : bodyTextOverflow,
            maxLines: bodyMaxLines == null ? _maxLines : bodyMaxLines,
          ),
        ),
      ),
    );
  }
}


lib/parser/web_page_parser.dart
/*import 'package:collection/collection.dart' show IterableExtension;*/

import 'package:html/dom.dart';
import 'package:html/dom.dart' hide Text;
import 'package:html/parser.dart' as parser;
import 'package:http/http.dart' as http;

class WebPageParser {
  static Future<Map> getData(String url) async {
    var response = await http.get(Uri.parse(url));

    return getDataFromResponse(response, url);
  }

  static Map<dynamic, dynamic> getDataFromResponse(
      http.Response response, String url) {
    var requiredAttributes = ['title', 'image'];
    var data = {};
    if (response.statusCode == 200) {
      var document = parser.parse(response.body);
      var openGraphMetaTags = _getOgPropertyData(document);

      openGraphMetaTags.forEach((element) {
        var ogTagTitle = element.attributes['property']!.split("og:")[1];
        var ogTagValue = element.attributes['content'];
        if ((ogTagValue != null && ogTagValue != "") ||
            requiredAttributes.contains(ogTagTitle)) {
          if (ogTagTitle == "image" && !ogTagValue!.startsWith("http")) {
            data[ogTagTitle] = "http://" + _extractHost(url) + ogTagValue;
          } else {
            data[ogTagTitle] = ogTagValue;
          }
        }
      });
      _scrapeDataToEmptyValue(data, document, url);
    }
    return data;
  }

  static String _extractHost(String link) {
    Uri uri = Uri.parse(link);
    return uri.host;
  }

  static void _scrapeDataToEmptyValue(Map data, Document document, String url) {
    if (!data.containsKey("title") ||
        data["title"] == null ||
        data["title"] == "") {
      data["title"] = _scrapeTitle(document);
    }

    if (!data.containsKey("image") ||
        data["image"] == null ||
        data["image"] == "") {
      data["image"] = _scrapeImage(document, url);
    }

    if (!data.containsKey("description") ||
        data["description"] == null ||
        data["description"] == "") {}
    data["description"] = _scrapeDescription(document);
  }

  static String _scrapeTitle(Document document) {
    var titleTags = document.head!.getElementsByTagName("title");
    if (titleTags.isNotEmpty && titleTags.first != null) {
      return titleTags.first.text;
    }
    return "";
  }

  static String _scrapeDescription(Document document) {
    var meta = document.getElementsByTagName("meta");
    String? description = "";
    var metaDescription =
        meta.firstWhere((e) => e.attributes["name"] == "description");

    if (metaDescription != null) {
      description = metaDescription.attributes["content"];
    }

    if (description != null && description != "") {
      return description;
    } else {
      description = document.head!.getElementsByTagName("title")[0].text;
    }
    return description;
  }

  static String? _scrapeImage(Document document, String url) {
    var images = document.body!.getElementsByTagName("img");
    String? imageSrc = "";
    if (images.length > 0) {
      imageSrc = images[0].attributes["src"];

      if (!imageSrc!.startsWith("http")) {
        imageSrc = "http://" + _extractHost(url) + imageSrc;
      }
    }
    if (imageSrc == "") {
      print("WARNING - WebPageParser - " + url);
      print(
          "WARNING - WebPageParser - image might be empty. Tag <img> was not found.");
    }
    return imageSrc;
  }

  static List<Element> _getOgPropertyData(Document document) {
    return document.head!.querySelectorAll("[property*='og:']");
  }

  static String? addWWWPrefixIfNotExists(String? uri) {
    if (uri == null || uri == "") {
      return uri;
    }

    Uri? prefixUri;
    Uri parsedUri = Uri.parse(uri);
    if (!parsedUri.host.startsWith('www')) {
      prefixUri = parsedUri.replace(host: 'www.' + parsedUri.host);
    }
    return prefixUri.toString();
  }
}


lib/link_previewer.dart
library link_previewer;

import 'package:url_launcher/url_launcher.dart';
import 'package:http/http.dart' as http;
import 'package:html/parser.dart' as parser;
import 'package:html/dom.dart' hide Text;
import 'package:flutter/material.dart' hide Element;
import 'dart:async';

part 'parser/web_page_parser.dart';

part 'package:link_previewer/horizontal_link_view.dart';

part 'package:link_previewer/vertical_link_preview.dart';

part 'package:link_previewer/content_direction.dart';

class LinkPreviewer extends StatefulWidget {
  LinkPreviewer({
    Key? key,
    required this.link,
    this.titleFontSize,
    this.bodyFontSize,
    this.backgroundColor = Colors.white,
    this.borderColor = Colors.deepOrangeAccent,
    this.titleTextColor = Colors.black,
    this.bodyTextColor = Colors.grey,
    this.defaultPlaceholderColor,
    this.borderRadius,
    this.placeholder,
    this.showTitle = true,
    this.showBody = true,
    this.direction = ContentDirection.horizontal,
    this.bodyTextOverflow,
    this.bodyMaxLines,
  })  : assert(link != null),
        super(key: key);

  final String link;
  final double? titleFontSize;
  final double? bodyFontSize;
  final Color backgroundColor;
  final Color borderColor;
  final Color? defaultPlaceholderColor;
  final Color titleTextColor;
  final Color bodyTextColor;
  final double? borderRadius;
  final ContentDirection direction;
  final Widget? placeholder;
  final bool showTitle;
  final bool showBody;
  final TextOverflow? bodyTextOverflow;
  final int? bodyMaxLines;

  @override
  _LinkPreviewer createState() => _LinkPreviewer();
}

class _LinkPreviewer extends State<LinkPreviewer> {
  Map? _metaData;
  double? _height;
  String? _link;
  Color? _placeholderColor;
  bool _failedToLoadImage = false;

  @override
  void initState() {
    super.initState();
    _link = widget.link.trim();
    if (_link!.startsWith("https")) {
      _link = "http" + _link!.split("https")[1];
    }
    _placeholderColor = widget.defaultPlaceholderColor == null
        ? Color.fromRGBO(235, 235, 235, 1.0)
        : widget.defaultPlaceholderColor;
    _fetchData();
  }

  double _computeHeight(double screenHeight) {
    if (widget.direction == ContentDirection.horizontal) {
      return screenHeight * 0.12;
    } else {
      return screenHeight * 0.25;
    }
  }

  void _fetchData() {
    if (!isValidUrl(_link)) {
      throw Exception("Invalid link");
    } else {
      _getMetaData(_link);
    }
  }

  void _validateImageUri(uri) {
    precacheImage(NetworkImage(uri), context, onError: (e, stackTrace) {
      setState(() {
        _failedToLoadImage = true;
      });
    });
  }

  String? _getUriWithPrefix(uri) {
    return WebPageParser._addWWWPrefixIfNotExists(uri);
  }

  void _getMetaData(link) async {
    Map data = await WebPageParser.getData(link);
    if (data != null) {
      _validateImageUri(data['image']);
      setState(() {
        _metaData = data;
      });
    } else {
      setState(() {
        _metaData = null;
      });
    }
  }

  bool isValidUrl(link) {
    String regexSource =
        "^(https?)://[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]";
    final regex = RegExp(regexSource);
    final matches = regex.allMatches(link);
    for (Match match in matches) {
      if (match.start == 0 && match.end == link.length) {
        return true;
      }
    }
    return false;
  }

  void _launchURL(url) async {
    if (await canLaunch(url)) {
      await launch(url);
    } else {
      throw 'Could not launch $url';
    }
  }

  @override
  Widget build(BuildContext context) {
    _height = _computeHeight(MediaQuery.of(context).size.height);

    return _metaData == null
        ? widget.placeholder == null
            ? _buildPlaceHolder(_placeholderColor, _height)
            : widget.placeholder!
        : _buildLinkContainer();
  }

  Widget _buildPlaceHolder(Color? color, double? defaultHeight) {
    return Container(
      height: defaultHeight,
      child: LayoutBuilder(builder: (context, constraints) {
        var layoutWidth = constraints.biggest.width;
        var layoutHeight = constraints.biggest.height;

        return Container(
          decoration: new BoxDecoration(
            color: color,
            border: Border.all(
              color: widget.borderColor == null
                  ? widget.backgroundColor
                  : widget.borderColor,
              width: widget.borderColor == null ? 0.0 : 1.0,
            ),
            borderRadius: BorderRadius.all(Radius.circular(
                widget.borderRadius == null ? 3.0 : widget.borderRadius!)),
          ),
          width: layoutWidth,
          height: layoutHeight,
        );
      }),
    );
  }

  Widget _buildLinkContainer() {
    return Container(
      decoration: new BoxDecoration(
        color: widget.backgroundColor,
        border: Border.all(
          color: widget.borderColor == null
              ? widget.backgroundColor
              : widget.borderColor,
          width: widget.borderColor == null ? 0.0 : 1.0,
        ),
        borderRadius: BorderRadius.all(Radius.circular(
            widget.borderRadius == null ? 3.0 : widget.borderRadius!)),
      ),
      height: _height,
      child: _buildLinkView(
          _link,
          _metaData!['title'] == null ? "" : _metaData!['title'],
          _metaData!['description'] == null ? "" : _metaData!['description'],
          _metaData!['image'] == null ? "" : _metaData!['image'],
          _launchURL,
          widget.showTitle,
          widget.showBody,
          widget.borderRadius == null ? 3.0 : widget.borderRadius),
    );
  }

  Widget _buildLinkView(link, title, description, imageUri, onTap, showTitle,
      showBody, borderRadius) {
    if (widget.direction == ContentDirection.horizontal) {
      return HorizontalLinkView(
        url: link,
        title: title,
        description: description,
        imageUri: _failedToLoadImage == false
            ? imageUri
            : _getUriWithPrefix(imageUri)!,
        onTap: onTap,
        showTitle: showTitle,
        showBody: showBody,
        bodyTextOverflow: widget.bodyTextOverflow,
        bodyMaxLines: widget.bodyMaxLines,
        titleTextColor: widget.titleTextColor,
        bodyTextColor: widget.bodyTextColor,
        borderRadius: borderRadius,
      );
    } else {
      return VerticalLinkPreview(
        url: link,
        title: title,
        description: description,
        imageUri: _failedToLoadImage == false
            ? imageUri
            : _getUriWithPrefix(imageUri)!,
        onTap: onTap,
        showTitle: showTitle,
        showBody: showBody,
        bodyTextOverflow: widget.bodyTextOverflow,
        bodyMaxLines: widget.bodyMaxLines,
        titleTextColor: widget.titleTextColor,
        bodyTextColor: widget.bodyTextColor,
        borderRadius: borderRadius,
      );
    }
  }
}
